rust/fastsim-core/src/vehicle_utils.rs

```diff
@@ -1771,6 +1773,99 @@ pub fn import_and_save_all_vehicles(
+    pub self_url: Option<String>,
+    pub git: Option<String>,
+    pub html: Option<String>,
+}
+
+#[derive(Deserialize)]
+pub struct GitObjectInfo {
+    pub name: String,
+    pub path: String,
+    pub sha: Option<String>,
+    pub size: Option<i64>,
+    pub url: String,
+    pub html_url: Option<String>,
+    pub git_url: Option<String>,
+    pub download_url: Option<String>,
+    #[serde(rename = "type")]
+    pub url_type: String,
+    #[serde(rename = "_links")]
+    pub links: Option<ObjectLinks>,
+}
+
+const VEHICLE_REPO_LIST_URL: &'static str =
+    &"https://api.github.com/repos/NREL/fastsim-vehicles/contents/public";
+
+/// Function that takes a url and calls the url. If a 503 or 429 error is
+/// thrown, it tries again after a pause, up to four times. It returns either a
+/// result or an error.  
+/// # Arguments  
+/// - url: url to be called
+/// Source: https://docs.rs/ureq/latest/ureq/enum.Error.html
+fn get_response<S: AsRef<str>>(url: S) -> Result<Response, OtherError> {
+    for _ in 1..4 {
+        match ureq::get(url.as_ref()).call() {
+            Err(Status(503, r)) | Err(Status(429, r)) | Err(Status(403, r)) => {
+                let retry: Option<u64> = r.header("retry-after").and_then(|h| h.parse().ok());
+                let retry = retry.unwrap_or(5);
+                eprintln!("{} for {}, retry in {}", r.status(), r.get_url(), retry);
+                thread::sleep(Duration::from_secs(retry));
+            }
+            result => return result,
+        };
+    }
+    // Ran out of retries; try one last time and return whatever result we get.
+    ureq::get(url.as_ref()).call()
+}
+
+/// Returns a list of vehicle file names in the Fastsim Vehicle Repo, or,
+/// optionally, a different GitHub repo, in which case the url provided needs to
+/// be the url for the file tree within GitHub for the root folder the Rust
+/// objects, for example
+/// "https://api.github.com/repos/NREL/fastsim-vehicles/contents/public"  
+/// Note: for each file, the output will list the vehicle file name, including
+/// the path from the root of the repository  
+/// # Arguments  
+/// - repo_url: url to the GitHub repository, Option, if None, defaults to the
+///   FASTSim Vehicle Repo
+pub fn fetch_github_list(repo_url: Option<String>) -> anyhow::Result<Vec<String>> {
+    let repo_url = repo_url.unwrap_or(VEHICLE_REPO_LIST_URL.to_string());
+    let response = get_response(repo_url)?.into_reader();
+    let github_list: Vec<GitObjectInfo> =
+        serde_json::from_reader(response).with_context(|| "Cannot parse github vehicle list.")?;
+    let mut vehicle_name_list: Vec<String> = Vec::new();
+    for object in github_list.iter() {
+        if object.url_type == "dir" {
+            let url = &object.url;
+            let vehicle_name_sublist = fetch_github_list(Some(url.to_owned()))?;
+            for name in vehicle_name_sublist.iter() {
+                vehicle_name_list.push(name.to_owned());
+            }
```

Comment:
@robinsteuteville Would this only look one level deep into directories?

Reply:
@kylecarow I believe it should look as many levels deep into directories as needed, since it calls fetch_github_list recursively within itself, which should mean that each time it comes across a subdirectory within the current directory it is in, it should start the process over again for that subdirectory, which in turn will uncover any subsubdirectories, etc. Does make my brain hurt a bit, so if you see a problem within the recursive logic let me know, I definitely could be thinking about it wrong!